lang/java/reef-utils/src/main/java/org/apache/reef/util/MultiAsyncToSync.java

```diff
@@ -0,0 +1,104 @@
+ * limitations under the License.
+ */
+package org.apache.reef.util;
+
+import org.apache.reef.util.exception.InvalidBlockedCallerIdentifierException;
+
+import java.util.ArrayDeque;
+import java.util.HashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Assists a class in presenting a synchronous interface that is implemented
+ * via asynchronous interfaces and events. When a method call is received
+ * by the interface, parameter values are captured and asynchronous processing
+ * started, the caller is put to sleep by calling block() with the internal
+ * interface caller identifier. After all of the asynchronous processing
+ * is complete the caller is released with a call to release().
+ */
+public final class MultiAsyncToSync {
+  private static final Logger LOG = Logger.getLogger(MultiAsyncToSync.class.getName());
+
+  private final ArrayDeque<SimpleCondition> freeQueue = new ArrayDeque<>();
+  private final HashMap<Long, SimpleCondition> sleeperMap = new HashMap<>();
+  private final long timeoutPeriod;
+  private final TimeUnit timeoutUnits;
+
+  /**
+   * Initialize a multiple asynchronous to synchronous object with a specified timeout value.
+   * @param timeoutPeriod The length of time in units given by the the timeoutUnits
+   *                      parameter before the condition automatically times out.
+   * @param timeoutUnits The unit of time for the timeoutPeriod parameter.
+   */
+  public MultiAsyncToSync(final long timeoutPeriod, final TimeUnit timeoutUnits) {
+    this.timeoutPeriod = timeoutPeriod;
+    this.timeoutUnits = timeoutUnits;
+  }
+
+  /**
+   * Put the caller to sleep on a specific release identifier.
+   * @param identifier The identifier required to awake the caller via the release() method.
+   * @return A boolean value that indicates whether or not a timeout occurred.
+   * @throws InterruptedException The thread was interrupted while waiting on a condition.
+   */
+  public boolean block(final long identifier) throws InterruptedException {
+    final SimpleCondition call;
+    synchronized (sleeperMap) {
+      // Get an condition variable to block the calling thread.
+      if (sleeperMap.containsKey(identifier)) {
+        throw new RuntimeException(String.format("Duplicate identifier [%d] in sleeper map", identifier));
+      }
+      if (freeQueue.isEmpty()) {
+        freeQueue.addLast(new SimpleCondition(timeoutPeriod, timeoutUnits));
+      }
+      call = freeQueue.getFirst();
+      sleeperMap.put(identifier, freeQueue.getFirst());
+    }
+
+    LOG.log(Level.INFO, "Putting caller to sleep on identifier [{0}]", identifier);
+    // Put the call to sleep until the ack comes back.
+    final boolean timeoutOccurred = call.waitForSignal();
+    if (timeoutOccurred) {
+      synchronized (sleeperMap) {
+        freeQueue.addLast(sleeperMap.remove(identifier));
```

Comment:
Lock order takes the call lock and then the container lock. The call lock protects across simultaneous block/release for the same identifier and the container lock protects against simultaneous block/release for the different identifiers.